### DISCUSSION TITLE
Cross compiling RPI with external libraries

### DISCUSSION THREAD

**Discussion by markospacetech** [nasa/fprime#1242] [Q&A] [answered]
Hello all,

I am looking to create a camera component utilizing the [raspicam ](https://github.com/rmsalinas/raspicam) library. 
In the component's CmakeLists.txt I have
`set(raspicam_DIR "/usr/local/lib/raspicam-0.1.9/build")`
`find_package(raspicam REQUIRED)`
`target_link_libraries(RPI_PiCamera PUBLIC ${raspicam_LIBS})`

which successfully builds the component, however when the build gets to
`[100%] Linking CXX executable bin/raspberrypi/RPI` it errors with

`/opt/rpi/tools/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf/bin/../lib/gcc/arm-linux-gnueabihf/4.9.3/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lraspicam`
'
Do I need to add the build (.so) path to some env variable like with the RPI build tools? Not sure if its that or something with my CMake config/linking so figured I'd ask here as well.

Cheers🚀

**Comment by LeStarch** [comment, marked as answer]
That looks like the right approach.  Did you check the architecture of the libraries in that raspicam directory?  Perhaps the libraries there are for another architecture?

**Reply by markospacetech** [reply]
Not sure what architecture means here. How do I know which one the library or fprime uses?

**Reply by LeStarch** [reply]
The tools you are building with is an older cross-compiler for the RPI.  Specifically it is building for 32bit RPI.  However, newer RPIs support arm64 (with a different cross compiler).  The steps I'd take:

1. Print out what libraries were found
```
set(raspicam_DIR "/usr/local/lib/raspicam-0.1.9/build")
find_package(raspicam REQUIRED)
message(STATUS "*** Libraries Here: ${raspicam_LIBS}")
target_link_libraries(RPI_PiCamera PUBLIC ${raspicam_LIBS})
```

This will let you know what libraries are being found in the package search.  Then run the file command to see what architecture and format these are:

For each library path:
```
file <path to library as shown in the above message>
```
This will tell you the format (probably ELF) and an architecutre (32bit 64bit arm).  Post the found libraries and the file call output here and perhaps we can help you debug some more!

**Reply by markospacetech** [reply]
Message raspicam_LIBS just printed "raspicam", however after running `file` in the build directory I get

```
raspicam/build/src/CMakeFiles:           directory

raspicam/build/src/cmake_install.cmake:  ASCII text

raspicam/build/src/libraspicam.so:       ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=bc4b48e73515423c87b33275228a41ce73411cc1, not stripped

raspicam/build/src/libraspicam.so.0.1:   ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=bc4b48e73515423c87b33275228a41ce73411cc1, not stripped

raspicam/build/src/libraspicam.so.0.1.9: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=bc4b48e73515423c87b33275228a41ce73411cc1, not stripped

raspicam/build/src/Makefile:             makefile script, ASCII text
```

**Reply by markospacetech** [reply]
That was for the build I compiled on the RPI. I also tried compiling from my linux machine (ubuntu 20) but I still got the same message during fprime-util build. The architecture is also different from the PI here.

```
libraspicam.so:       symbolic link to libraspicam.so.0.1
libraspicam.so.0.1:   symbolic link to libraspicam.so.0.1.9
libraspicam.so.0.1.9: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=5c26f8138688d5a5cb010f2a0a78bd540c699dae, not stripped
```

**Reply by LeStarch** [reply]
```
target_link_libraries(RPI_PiCamera PUBLIC ${raspicam_DIR}/src/libraspicam.so) 
```
Should work.  

`find_package` will only work is raspicam is already a CMake package.

**Reply by markospacetech** [reply]
That did the trick! I also needed to make sure I was using the same fprime cross compile tools at RPI_TOOLCHAIN_DIR with the `-std=gnu++11` c flag to compile the libraries.

Huge thanks!!

**Reply by LeStarch** [reply]
Glad we could help!
